### PR TITLE
Ensure newly paired devices are in storage before completing pairing

### PR DESCRIPTION
**Pull request by sagar-apple**
#### Problem
There's a race between persistent storage and the `OnPairingComplete` callback.

Now that network provisioning has been separated from the rendezvous flow, it has exposed a bug with how pairing indications are sent back to the caller. 

When the Rendezvous session is completed, the DeviceComissioner sends a "Pairing Complete" indication to the caller. However because the DeviceComissioner(and Controller for that matter) is always using async storage APIs, it's possible that the device is not actually in storage by the time the client is informed about the pairing completion. 

 #### Summary of Changes
- Add a sync Set Storage API for String based values.
- Add optional sync flags to the storage helpers used inside `CHIPDeviceController` to allow sync writes to storage when needed.
- Updated how Rendezvous session stores the paired device.

### CODE REVIEW

**Review by vivien-apple — APPROVED** [review]

**Comment by vivien-apple** [comment]
Seems to conflicts a bit with #6286.

**Comment by sagar-apple** [comment]
> Seems to conflicts a bit with #6286.

Yup either one of us can rebase. Nbd.
With Andrei's changes my pr just becomes a reordering of calls to storage.

**Review by msandstedt — APPROVED** [review]

**Comment by sagar-apple** [comment]
paging @andy31415 @bzbarsky-apple @chrisdecenzo @Damian-Nordic @hawk248 @jepenven-silabs since this was pushed over the weekend and might have gotten buried

**Review by bzbarsky-apple — COMMENTED** [review]
Please coordinate with #6286 as to which approach we want to take here.

**Review by bzbarsky-apple — APPROVED** [review]

**Comment by sagar-apple** [comment]
@andreilitvin think this can go in since #6286 needs a little more work? I know we'll end up undoing a lot of this but it fixes some of the racey stuff in the pairing flow.

File: src/controller/CHIPDeviceController.cpp [hunk 1]
```diff
@@ -926,7 +934,12 @@ void DeviceCommissioner::OnRendezvousComplete()
     mPairedDevices.Insert(device->GetDeviceId());
     mPairedDevicesUpdated = true;
 
-    PersistDevice(device);
+    // Newly paired devices should be stored synchronously because we cannot mark pairing complete
+    // until they are available in storage
+    PersistDevice(device, true);
```
Comment (andy31415) [inline src/controller/CHIPDeviceController.cpp:939]: nit: boolean arguments are non-obvious in general. One option would be to `PersitDevice(device, true /* sync */)` however even that feels awkward.

I would suggest using an enum of `kPersistSync` and `kPersistAsync` so that the `PersistDeviceList(kPersistSync)` is immediately obvious and type safe.

Comment (sagar-apple) [reply src/controller/CHIPDeviceController.cpp:939]: I will add an enum here. Thanks!

**Review by mspang — COMMENTED** [review]
If we move to synchronous storage APIs (or even just use them, even occasionally) then threads that care deeply about their responsiveness will not be able to run CHIP.

Are we really OK taking that step?

Async is definitely more complicated, but storage is one of those things that can be so slow that it's a deal breaker for threads with latency constraints.

**Comment by sagar-apple** [comment]
> If we move to synchronous storage APIs (or even just use them, even occasionally) then threads that care deeply about their responsiveness will not be able to run CHIP.
> 
> Are we really OK taking that step?
> 
> Async is definitely more complicated, but storage is one of those things that can be so slow that it's a deal breaker for threads with latency constraints.

That's fair, but this PR doesn't move everything to the Sync API. 
It only forces a couple of writes to be Sync because it's critical to have that state settled before we can complete our callback. Sounds like more of a review for #6286?

**Comment by mspang** [comment]
> > If we move to synchronous storage APIs (or even just use them, even occasionally) then threads that care deeply about their responsiveness will not be able to run CHIP.
> > Are we really OK taking that step?
> > Async is definitely more complicated, but storage is one of those things that can be so slow that it's a deal breaker for threads with latency constraints.
> 
> That's fair, but this PR doesn't move everything to the Sync API.
> It only forces a couple of writes to be Sync because it's critical to have that state settled before we can complete our callback. Sounds like more of a review for #6286?

What I'm saying is if it's even possible for CHIP to make initiate a synchronous storage operation in steady state operation then applications will be forced to move CHIP off of their latency critical threads. Not all applications, but certainly the applications that pay close attention to latency will. Which is a shame, because CHIP is a protocol stack and not something that fundamentally requires blocking I/O.

It's a preexisting issue since we already have synchronous reads. This PR takes a step further down the synchronous path, and Andrei's goes all the way.

**Comment by sagar-apple** [comment]
> What I'm saying is if it's even possible for CHIP to make initiate a synchronous storage operation in steady state operation then applications will be forced to move CHIP off of their latency critical threads. Not all applications, but certainly the applications that pay close attention to latency will. Which is a shame, because CHIP is a protocol stack and not something that fundamentally requires blocking I/O.
> 

Yeah I totally agree. But fwiw, the Darwin SDK already keeps CHIP on its own thread for that reason but I'm not sure if the other controllers do that as well.

**Comment by woody-apple** [comment]
@saurabhst @hawk248 ?

**Comment by andy31415** [comment]
> If we move to synchronous storage APIs (or even just use them, even occasionally) then threads that care deeply about their responsiveness will not be able to run CHIP.
> 
> Are we really OK taking that step?
> 
> Async is definitely more complicated, but storage is one of those things that can be so slow that it's a deal breaker for threads with latency constraints.

I would be open to any approach: either fully sync or fully async, as long as there is only one and is consistently used.
I believe both can be moved back & forth (either use an IO thread or have blocking waits).

The way I got to the 'full sync'
   - KVS currently only has sync API (@rgoliver - any feedback on design choices? Why did we not start as async?)
   - Existing async facilities were not really used within chip - callbacks were empty, reads were not used, delete success ignored etc.
  
 Once everything has a single interface (KVS or otherwise) we can choose to move things around.

File: src/darwin/CHIPTool/CHIPTool/View Controllers/DeviceSelector.m [hunk 2]
```diff
@@ -43,7 +43,7 @@ - (void)refreshDeviceList
 {
     uint64_t nextDeviceID = CHIPGetNextAvailableDeviceID();
     _deviceList = [NSMutableArray new];
-    for (uint64_t i = 0; i < nextDeviceID; i++) {
+    for (uint64_t i = 1; i < nextDeviceID; i++) {
```
Comment (andy31415) [inline src/darwin/CHIPTool/CHIPTool/View Controllers/DeviceSelector.m:46]: why the change in indexing here? Seems unrelated to persistence logic ... is it?

C/C++ array indexing from 1 is usually odd. Could you add a comment explaining?

File: src/lib/core/CHIPPersistentStorageDelegate.h [hunk 3]
```diff
@@ -126,6 +126,18 @@ class DLL_EXPORT PersistentStorageDelegate
...
+     * @brief
+     *   This is a synchronous Set API.
+     *   This API should be used sparingly, since it may block for some duration.
+     *
+     *   Set the value for the key to a null terminated string.
+     *
+     * @param[in] key Key to be set
+     * @param[in] value Value to be set
+     */
+    virtual CHIP_ERROR SyncSetKeyValue(const char * key, const char * value) { return CHIP_ERROR_NOT_IMPLEMENTED; }
```
Comment (andy31415) [inline src/lib/core/CHIPPersistentStorageDelegate.h:139]: Could we use and implement the void* value instead?

I find that we often need to serialize binary data: all the admin info and device info are binary and we force base64 encoding/decoding everywhere. It seems more convenient to have binary support only and interfaces for that already exist.

I think this proliferates more API split: we have sync/async, we have char * and void * which are likely not interoperable (null terminator handling and data format in general, it seems some implementation implement some and not the rest). This is the problem I was trying to solve over many PRs and it seems to get only worse. In this case we add a 'SyncSetKeyValue' on char * that is only implemented on Darwin and will not exist on many other platforms

**Review by andy31415 — CHANGES_REQUESTED** [review]
Marking changes requested: I find that we are adding more and more disperse APIs that are not implemented by all delegates and this goes backwards towards having a consistent storage method.

I would go fo:
   - use void *
   - if not using void *, then make the method pure virtual and force implement everywhere (it will be very painful!) so that we stop having diverging api capabilities here. It seems that the DeviceController now expects that the sync setting exists everywhere.

File: src/darwin/CHIPTool/CHIPTool/View Controllers/DeviceSelector.m (hunk 2 above)
Comment (sagar-apple) [reply src/darwin/CHIPTool/CHIPTool/View Controllers/DeviceSelector.m:46]: Yeah I can do that. Node ID 0 is reserved and so there's no sense is looking for a device with that ID. I should have left a comment.

File: src/lib/core/CHIPPersistentStorageDelegate.h (hunk 3 above)
Comment (sagar-apple) [reply src/lib/core/CHIPPersistentStorageDelegate.h:139]: We need another PR to start storing thing as void * and len instead of char *. I effectively hotfixed this without adding too much functional logic to the existing code. Your points are valid though.

**Comment by msandstedt** [comment]
> > What I'm saying is if it's even possible for CHIP to make initiate a synchronous storage operation in steady state operation then applications will be forced to move CHIP off of their latency critical threads. Not all applications, but certainly the applications that pay close attention to latency will. Which is a shame, because CHIP is a protocol stack and not something that fundamentally requires blocking I/O.
> 
> Yeah I totally agree. But fwiw, the Darwin SDK already keeps CHIP on its own thread for that reason but I'm not sure if the other controllers do that as well.

I think this is a likely design pattern, at least for posix, darwin and non-embedded.  Even if an existing application _could_ be driven from the CHIP event loop, many will choose not to rework their applications to do this, or will not trust CHIP not to block (which is apparently wise).

Having said that, I echo the concern that this is not a positive direction.  Async `Set` and `Delete` were easy; async `Get` is slightly more difficult to do in a useful way (and was never more than a stub in the controller) and so was axed.  Now we have the problem of needing to sequence around the synchronous Get.

Why don't we just have an async queue of I/O transactions?  I doubt anybody will disagree that this is the right approach.  Also, if you are crossing a thread boundary and supporting persistence in an existing event loop context, and your I/O primitives _are_ async, this starts to look really silly.  You have to stitch a dispatch and callback back together inside the blocking CHIP context.  That's not terribly difficult, but it is unfortunate.

**Comment by msandstedt** [comment]
> * Existing async facilities were not really used within chip - callbacks were empty, reads were not used, delete success ignored etc.

This is fair.  The original async storage delegate Get interface was indeed useless.

File: src/lib/core/CHIPPersistentStorageDelegate.h (hunk 3 above)
Comment (sagar-apple) [reply src/lib/core/CHIPPersistentStorageDelegate.h:139]: Looking at this a bit more. The Darwin APIs are all (Char *) based. To do this using the void * API, the darwin delegate implmentation needs to be rewritten to use `NSData` or generic pointers `id` (probably NSData?). I'm currently not planning to do that in this PR. Which means this is kinda dead in the water until we change the darwin APIs first.

**Comment by sagar-apple** [comment]
This PR clearly needs more work that might be outside its own scope. At this time, to avoid duplicating with Andrei's work, I'm going to have to park this until all the storage APIs switch to using buffers + len rather than chars. + we need some mechanism to sync or chain calls (via callbacks). Converting it to a draft.
